tests/languages/download_test.py

```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
+
+import platform
+import stat
+from http.server import BaseHTTPRequestHandler
+from http.server import HTTPServer
+from os import chmod
+from pathlib import Path
+from threading import Thread
+
+from pytest import fixture
+from pytest import TempdirFactory
+
+import pre_commit.constants as C
+from pre_commit import lang_base
+from pre_commit.languages import download
+from pre_commit.languages.download import ChecksumMismatchError
+from pre_commit.languages.download import ENVIRONMENT_DIR
+from pre_commit.languages.download import get_default_version
+from pre_commit.languages.download import SRI
+from pre_commit.prefix import Prefix
+from testing.language_helpers import run_language
+
+shellscript = b'#!/bin/sh\necho hello\nexit 123'
+shell_checksum = SRI('sha256-oRJkj6Cr8nWIivZ9d3W+rVZt/aSW1l9YtxSVh+GtIHM=')
+batch_script = b'@echo off\necho hello\nexit 123'
+batch_checksum = SRI('sha256-L63Nefq+fKVIm24IKqlcqbJmc1rrJD3dKhIvutFK+IA=')
+if platform.system() == 'Windows':
+    correct_checksum = batch_checksum
+else:
+    correct_checksum = shell_checksum
+
+healthy_dependencies = [
+    f"""linux/amd64
+{shell_checksum}
+http://127.0.0.1:5555
+test.bat""",
+    f"""windows/amd64
+{batch_checksum}
+http://127.0.0.1:5555
+test.bat""",
+]
+
+deps_invalid_length = [
+    """linux/amd64
+sha256-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+http://127.0.0.1:5555
+test.bat""",
+    """windows/amd64
+sha256-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+http://127.0.0.1:5555
+test.bat""",
+]
+
+deps_wrong_checksum = [
+    """linux/amd64
+sha256-oRJkj6Cr8nWIivZ9d4W+rVZt/aSW1l9YtxSVh+GtIHM=
+http://127.0.0.1:5555
+test.bat""",
+    """windows/amd64
+sha256-L63Nefq+fKVIm25IKqlcqbJmc1rrJD3dKhIvutFK+IA=
+http://127.0.0.1:5555
+test.bat""",
+]
+
+
+class HTTPRequestHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        self.send_response(200)
+        self.end_headers()
+        if platform.system() == 'Windows':
+            self.wfile.write(batch_script)
+        else:
+            self.wfile.write(shellscript)
+
+
+class server():
+    def __init__(self):
+        self.httpd = HTTPServer(('localhost', 5555), HTTPRequestHandler)
+        self.thread = Thread(target=self.httpd.serve_forever)
+
+    def start(self):
+        self.thread.start()
+
+    def stop(self):
+        self.httpd.shutdown()
+        self.thread.join()
+
+
+@fixture
+def prefix(tmpdir: TempdirFactory) -> Prefix:
+    with tmpdir.as_cwd():
+        prefix = tmpdir.join('prefix').ensure_dir()
+        prefix = Prefix(str(prefix))
+        return prefix
```

Comment:
This needs to be a `yield` otherwise the `as_cwd()` context manager (`with`) will close out.

Reply:
Done